src/merkleblock.cpp

```diff
@@ -48,10 +23,14 @@ CMerkleBlock::CMerkleBlock(const CBlock& block, const std::set<uint256>& txids)
     for (unsigned int i = 0; i < block.vtx.size(); i++)
     {
         const uint256& hash = block.vtx[i]->GetHash();
-        if (txids.count(hash))
+        if (txids != nullptr && (*txids).count(hash)) {
             vMatch.push_back(true);
-        else
+        } else if (filter != nullptr && (*filter).IsRelevantAndUpdate(*block.vtx[i])) {
+            vMatch.push_back(true);
+            vMatchedTxn.push_back(std::make_pair(i, hash));
```

Comment:
You can avoid constructing an intermediate `std::pair<unsigned int, uint256>`, by using:

```c++
vMatchedTxn.emplace_back(i, hash);
```